lib/memoryprofiler/src/trait.rs

```diff
@@ -0,0 +1,169 @@
+pub trait MemoryUsage {
+    fn size_of(&self) -> usize;
+}
+
+/// Primitive types
+impl MemoryUsage for bool {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<bool>()
+    }
+}
+impl MemoryUsage for char {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<char>()
+    }
+}
+impl MemoryUsage for f32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f32>()
+    }
+}
+impl MemoryUsage for f64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<f64>()
+    }
+}
+impl MemoryUsage for i8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i8>()
+    }
+}
+impl MemoryUsage for i16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i16>()
+    }
+}
+impl MemoryUsage for i32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i32>()
+    }
+}
+impl MemoryUsage for i64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<i64>()
+    }
+}
+impl MemoryUsage for isize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<isize>()
+    }
+}
+impl MemoryUsage for u8 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u8>()
+    }
+}
+impl MemoryUsage for u16 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u16>()
+    }
+}
+impl MemoryUsage for u32 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u32>()
+    }
+}
+impl MemoryUsage for u64 {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<u64>()
+    }
+}
+impl MemoryUsage for usize {
+    fn size_of(&self) -> usize {
+        std::mem::size_of::<usize>()
+    }
+}
+
+// pointer
+// Pointers aren't necessarily safe to dereference, even if they're nonnull.
+
+// references
+impl<T: MemoryUsage> MemoryUsage for &T {
+    fn size_of(&self) -> usize {
+        (**self).size_of()
```

Comment:
I believe we also want to account for the memory usage of the `&` itself, which would be `usize`

Reply:
I've replaced this with:
```
    fn size_of(&self) -> usize {
        std::mem::size_of::<&T>()
    }
```
because if we visit the referred-to object then it may be double-counted with anyone else holding the immutable reference. For mutable references the same logic applies even though there can only be one mutable reference it still has to be referring to an object which is owned by somebody elsewhere.